### PR TITLE
add `selector-nesting-notation` rule

### LINKED ISSUE #8161: Add `selector-explicit-nesting`
### What is the problem you're trying to solve?

Require or disallow explicit [nesting selector](https://drafts.csswg.org/css-nesting-1/#nest-selector).

```css
a {
  & > b {}
}
```

or 

```css
a {
  > b {}
}
```

### What solution would you like to see?

A new rule:

- Name: `selector-explicit-nesting`
- Description: Require or disallow explicit nesting selectors
- Messages: 
   - `expected` = Expected explicit nesting selector
   - `rejected` = Unexpected explicit nesting selector
- Primary option: `"always"|"never"`
- Secondary options: none
- Category: Enforce conventions > Allowed, disallowed & required > Nesting
- Autofix?: Yes 
- Included in sharable configs?: Yes, `"always"`

### PR DESCRIPTION
**Pull request by sw1tch3roo**
> Which issue, if any, is this issue related to?

Closes #8161.

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

### CODE REVIEW

File: docs/user-guide/rules.md [hunk 1]
```diff
@@ -360,6 +360,7 @@ Enforce one representation of things that have multiple with these `notation` (s
 | [`keyframe-selector-notation`](../../lib/rules/keyframe-selector-notation/README.md)<br/>Specify keyword or percentage notation for keyframe selectors. | ✅ | 🔧 |
 | [`lightness-notation`](../../lib/rules/lightness-notation/README.md)<br/>Specify number or percentage notation for lightness. | ✅ | 🔧 |
 | [`media-feature-range-notation`](../../lib/rules/media-feature-range-notation/README.md)<br/>Specify context or prefix notation for media feature ranges. | ✅ | 🔧 |
+| [`selector-nesting-notation`](../../lib/rules/selector-nesting-notation/README.md)<br/>Specify explicit or implicit notation for nesting selectors. | ✅ | 🔧 |
```
Comment (Mouvedia) [inline docs/user-guide/rules.md]: Putting it in standard would be very opinionated.
It shouldn't be part of it.
e.g. lots of Sass codebases use `implicit` i.e. it would hinder the transition

Comment (sw1tch3roo) [reply docs/user-guide/rules.md]: > Putting it in standard would be very opinionated. It shouldn't be part of it. e.g. lots of Sass codebases use `implicit` i.e. it would hinder the transition

https://github.com/stylelint/stylelint/issues/8161#issuecomment-2541464378

maybe I don't understand something, but @jeddy3 pointed out: 
> Included in sharable configs?: Yes, in standard with "explicit"

Comment (Mouvedia) [reply docs/user-guide/rules.md]: Not everything is set in stone by the comments of the issue.
Some things change while reviewing the PR.

Comment (sw1tch3roo) [reply docs/user-guide/rules.md]: I get your point — `explicit` is definitely opinionated, and sass codebases using implicit nesting would hit a lot of friction.
Maybe the middle ground is:
- Keep it out of standard for now (or default to `implicit`?)
- Provide a separate preset for projects moving to native CSS nesting with `explicit` enforced

That way teams migrating from Sass aren’t blocked, and those aiming for modern CSS conventions can opt in easily.

Comment (Mouvedia) [reply docs/user-guide/rules.md]: > What do you mean?

Transition from a CSS preprocessor to CSS nesting.

File: lib/rules/selector-nesting-notation/README.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
+# selector-nesting-notation
+
+Enforce consistent usage of nesting selectors.
```
Comment (jeddy3) [inline lib/rules/selector-nesting-notation/README.md]:
```suggestion
Specify explicit or implicit notation for nesting selectors.
```

File: lib/rules/selector-nesting-notation/README.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
+# selector-nesting-notation
```
Comment (jeddy3) [inline lib/rules/selector-nesting-notation/README.md]: Shall we rename this `relative-selector-nesting-notation` as it only scoped to _relative selectors_ ([spec](https://drafts.csswg.org/css-nesting/#nesting)). I may have misunderstood the terminology in the spec, though.

It should make the descriptions of the options easier to keep consistent with other rules.

File: lib/rules/selector-nesting-notation/README.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+### `"explicit"`
+
+```json
+{
+  "selector-nesting-notation": "explicit"
+}
+```
+
+Require explicit nesting selector (`&`) when using combinators.
```
Comment (jeddy3) [inline lib/rules/selector-nesting-notation/README.md]:
```suggestion
Relative selectors _must always_ use explicit nesting selector notation.
```

Consistent with other notation rules, e.g. https://stylelint.io/user-guide/rules/alpha-value-notation

File: lib/rules/selector-nesting-notation/index.mjs [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+		});
+
+		if (!validOptions) return;
+
+		root.walkRules((ruleNode) => {
+			if (!isStandardSyntaxRule(ruleNode)) return;
+
+			if (!ruleNode.parent || !isRule(ruleNode.parent)) return;
+
+			const ruleSelector = getRuleSelector(ruleNode);
```
Comment (jeddy3) [inline lib/rules/selector-nesting-notation/index.mjs]: Let's account for [nested group rules](https://drafts.csswg.org/css-nesting/#conditionals):

```css
a { @media all { a {} } }
```

File: lib/rules/selector-nesting-notation/index.mjs [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+import { isRule } from '../../utils/typeGuards.mjs';
+import isStandardSyntaxRule from '../../utils/isStandardSyntaxRule.mjs';
+import report from '../../utils/report.mjs';
+import ruleMessages from '../../utils/ruleMessages.mjs';
+import validateOptions from '../../utils/validateOptions.mjs';
+
+const ruleName = 'selector-nesting-notation';
+
+const messages = ruleMessages(ruleName, {
+	expected: (unfixed, fixed) => `Expected "${unfixed}" to be "${fixed}"`,
```
Comment (jeddy3) [inline lib/rules/selector-nesting-notation/index.mjs]: Let's use `primary` as `fixed` and `unfixed` could be long strings ([conventions](https://stylelint.io/developer-guide/rules#add-problem-messages)).

Sorry, I got this wrong in my blueprint for the rule.

File: lib/rules/selector-nesting-notation/index.mjs [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+			selectorText = `& ${selectorText}`;
+		} else if (violation?.type === 'remove-nesting') {
+			// Remove & and following space from selector
+			selectorText = selectorText.substring(selectorText.startsWith('& ') ? 2 : 1);
+		}
+
+		// Fix spacing issues from CSS Tree generate()
+		// CSSTree sometimes omits spaces after combinators (>, +, ~) in generated output
+		// This regex adds the required space to ensure valid CSS formatting
+		selectorText = selectorText.replace(/([>+~])([a-z#.])/gi, '$1 $2');
```
Comment (jeddy3) [inline lib/rules/selector-nesting-notation/index.mjs]: I don't believe we can use CSSTree because it'll strip all comments and insignificant whitespace. Your workaround is pointing to this limitation.

We can use the PostCSS selector parser instead, and transform its AST to maintain comments and whitespace.

File: lib/rules/selector-nesting-notation/index.mjs [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+ * @returns {boolean}
+ */
+function hasNestingSelector(selector) {
+	let hasNesting = false;
+
+	selector.children.forEach((child) => {
+		if (child.type === 'NestingSelector') {
+			hasNesting = true;
+		}
+	});
```
Comment (jeddy3) [inline lib/rules/selector-nesting-notation/index.mjs]: Let's account for the likes of:

```css
a { :not(&) {} }
```

**Review by jeddy3 — CHANGES_REQUESTED** [review]
@sw1tch3roo Thanks for starting on this!

When a rule autofixes, we should consider non-significant whitespace and comments.  See the [commonly overlooked cases](https://stylelint.io/developer-guide/rules#commonly-overlooked-edge-cases) in the developer guide.

I've highlighted a few other edge cases in comments below.

**Comment by jeddy3** [comment]
Here's a version of the rule that uses the PostCSS Selector parser to account for some of the cases I've highlighted above, especially around preserving comments and whitespace. It simplifies a few things, but adds complexity back in to manage whitespace and comments in cases like:

```css
a { > .foo, /* bar */ + .baz {} }
```

Managing whitespace and comments is typically a complex part of a rule. I used `selector-not-notation` as a blueprint, which also dedicates a chunk of code towards this.

There are likely more edge cases to uncover and opportunities to refactor, but this should give you a starting point and direction when using the PostCSS selector parser. Once it's dropped in, it'd be good to look for more edge cases and/or shoring up the test coverage, as it's a rough cut:

<details><summary>Rule</summary>

```js
import parser from 'postcss-selector-parser';

import findNodeUpToRoot from '../../utils/findNodeUpToRoot.mjs';
import getRuleSelector from '../../utils/getRuleSelector.mjs';
import isStandardSyntaxRule from '../../utils/isStandardSyntaxRule.mjs';
import parseSelector from '../../utils/parseSelector.mjs';
import report from '../../utils/report.mjs';
import ruleMessages from '../../utils/ruleMessages.mjs';
import validateOptions from '../../utils/validateOptions.mjs';

const ruleName = 'selector-nesting-notation';

const messages = ruleMessages(ruleName, {
	expected: (primary) => `Expected ${primary} nesting selector notation`,
});

const meta = {
	url: 'https://stylelint.io/user-guide/rules/selector-nesting-notation',
	fixable: true,
};

/** @type {import('stylelint').CoreRules[ruleName]} */
const rule = (primary) => {
	return (root, result) => {
		const validOptions = validateOptions(result, ruleName, {
			actual: primary,
			possible: ['explicit', 'implicit'],
		});

		if (!validOptions) return;

		root.walkRules((ruleNode) => {
			if (!isStandardSyntaxRule(ruleNode)) return;

			if (!isNestedStyleRule(ruleNode)) return;

			const selectorRoot = parseSelector(getRuleSelector(ruleNode), result, ruleNode);

			if (!selectorRoot) return;

			selectorRoot.each((selector) => {
				if (selector.type !== 'selector') return;

				const hasNesting = isNestContaining(selector);
				const startsWithComb = startsWithCombinator(selector);
				const startsWithNesting = canUseImplicitNesting(selector);

				let fix;

				if (primary === 'explicit') {
					if (hasNesting && !startsWithComb) return;

					fix = () => fixExplicit(selector, ruleNode, selectorRoot);
				}

				if (primary === 'implicit') {
					if (startsWithComb || !startsWithNesting) return;

					fix = () => fixImplicit(selector, ruleNode, selectorRoot);
				}

				const index = selector.sourceIndex;
				const endIndex = index + selector.toString().trim().length;

				report({
					message: messages.expected,
					messageArgs: [primary],
					node: ruleNode,
					index,
					endIndex,
					ruleName,
					result,
					fix: {
						apply: fix,
						node: ruleNode,
					},
				});
			});
		});
	};
};

/**
 * Check if a rule is a nested style rule
 *
 * @param {import('postcss').Rule} ruleNode
 * @returns {boolean}
 */
function isNestedStyleRule(ruleNode) {
	return Boolean(findNodeUpToRoot(ruleNode, ({ type }) => type === 'rule'));
}

/**
 * Check if a selector contains the nesting selector
 *
 * @param {import('postcss-selector-parser').Selector} selector
 * @returns {boolean}
 */
function isNestContaining(selector) {
	let hasNestingSelector = false;

	selector.walkNesting(() => (hasNestingSelector = true));

	return hasNestingSelector;
}

/**
 * Check if a selector starts with a combinator
 *
 * @param {import('postcss-selector-parser').Selector} selector
 * @returns {boolean}
 */
function startsWithCombinator(selector) {
	const firstSignificantNode = selector.nodes.find((node) => node.type !== 'comment');

	return firstSignificantNode?.type === 'combinator';
}

/**
 * Check if a selector can use implicit nesting, i.e.
 * has a combinator after any comments or nesting selectors
 *
 * @param {import('postcss-selector-parser').Selector} selector
 * @returns {boolean}
 */
function canUseImplicitNesting(selector) {
	const foundNode = selector.nodes.find(
		(node) => node.type !== 'comment' && node.type !== 'nesting',
	);

	return foundNode?.type === 'combinator';
}

/**
 * Fix selector to use explicit nesting notation
 *
 * @param {import('postcss-selector-parser').Selector} selector
 * @param {import('postcss').Rule} ruleNode
 * @param {import('postcss-selector-parser').Root} selectorRoot
 */
function fixExplicit(selector, ruleNode, selectorRoot) {
	const firstSignificantNode = selector.nodes.find(
		(node) => !['comment', 'selector', 'root'].includes(node.type),
	);

	if (firstSignificantNode) {
		const { before } = firstSignificantNode.spaces;

		selector.insertBefore(
			firstSignificantNode,
			parser.nesting({
				spaces: {
					after: ' ',
					before,
				},
			}),
		);
		firstSignificantNode.spaces.before = '';
	}

	ruleNode.selector = selectorRoot.toString();
}

/**
 * Fix selector to use implicit nesting notation
 *
 * @param {import('postcss-selector-parser').Selector} selector
 * @param {import('postcss').Rule} ruleNode
 * @param {import('postcss-selector-parser').Root} selectorRoot
 */
function fixImplicit(selector, ruleNode, selectorRoot) {
	const firstNestingNode = selector.nodes.find((node) => node.type === 'nesting');
	let spaceBeforeFirstNestingNode = '';

	// Remove the first nesting node and store its spaces.before
	if (firstNestingNode) {
		spaceBeforeFirstNestingNode = firstNestingNode.spaces.before;
		firstNestingNode.remove();
	}

	const firstCombinatorNode = selector.nodes.find((node) => node.type === 'combinator');

	// Remove the first combinator node and update the next node's spaces
	if (firstCombinatorNode) {
		const nextNode = firstCombinatorNode.next();

		if (nextNode) {
			nextNode.spaces.before = spaceBeforeFirstNestingNode;
		}

		// If it's a descendant combinator remove it, otherwise clear its before space
		if (firstCombinatorNode.value === ' ') {
			firstCombinatorNode.remove();
		} else {
			firstCombinatorNode.spaces.before = '';
		}
	}

	ruleNode.selector = selectorRoot.toString();
}

rule.ruleName = ruleName;
rule.messages = messages;
rule.meta = meta;
export default rule;
```
</details>

And here some revised tests that:

1. adhere to our conventions
2. have testcase for some of the cases I've mentioned

<details><summary>Tests:</summary>

```js
import { stripIndent } from 'common-tags';

import rule from '../index.mjs';
const { messages, ruleName } = rule;

testRule({
	ruleName,
	config: ['explicit'],
	fix: true,
	computeEditInfo: true,

	accept: [
		{
			code: 'a {}',
			description: 'non-nested rule',
		},
		{
			code: '@media all { a {} }',
			description: 'rule in at-rule',
		},
		{
			code: 'a { & a {} }',
			description: 'explicit nesting with descendant combinator',
		},
		{
			code: 'a { & > a {} }',
			description: 'explicit nesting with child combinator',
		},
		{
			code: 'a { &:hover {} }',
			description: 'nesting with pseudo-class',
		},
		{
			code: 'a { &.foo {} }',
			description: 'nesting with class',
		},
		{
			code: 'a { :not(&) {} }',
			description: 'nesting in pseudo-class',
		},
		{
			code: 'a { a & {} }',
			description: 'nesting selector as parent',
		},
		{
			code: 'a { & /* foo */ a, .bar /* baz */ & {} }',
			description: 'nesting selector in list with comment',
		},
		{
			code: 'a { /* bar */ & a, /* baz */ .foo & {} }',
			description: 'nesting selector in list with starting comments',
		},
		{
			code: stripIndent`
				.a {
					& .foo {
						& > #bar {}
					}
				}
			`,
			description: 'multiple levels of explicit nesting',
		},
	],

	reject: [
		{
			code: 'a { a {} }',
			fixed: 'a { & a {} }',
			message: messages.expected('explicit'),
			line: 1,
			column: 5,
			endLine: 1,
			endColumn: 6,
			fix: {
				range: [4, 5],
				text: '& a',
			},
		},
		{
			code: 'a { > a {} }',
			fixed: 'a { & > a {} }',
			message: messages.expected('explicit'),
			line: 1,
			column: 5,
			endLine: 1,
			endColumn: 8,
			fix: {
				range: [4, 5],
				text: '& >',
			},
		},
		{
			code: 'a { [foo] {} }',
			fixed: 'a { & [foo] {} }',
			message: messages.expected('explicit'),
			line: 1,
			column: 5,
			endLine: 1,
			endColumn: 10,
			fix: {
				range: [4, 5],
				text: '& [',
			},
		},
		{
			code: 'a { + a & {} }',
			fixed: 'a { & + a & {} }',
			description: 'nesting as parent and implicit inital & before combinator',
			message: messages.expected('explicit'),
			line: 1,
			column: 5,
			endLine: 1,
			endColumn: 10,
			fix: {
				range: [4, 5],
				text: '& +',
			},
		},
		{
			code: 'a { @media all { a {} } }',
			fixed: 'a { @media all { & a {} } }',
			description: 'nesting in media query',
			message: messages.expected('explicit'),
			line: 1,
			column: 18,
			endLine: 1,
			endColumn: 19,
			fix: {
				range: [17, 18],
				text: '& a',
			},
		},
		{
			code: 'a { ~ /* foo */ a {} }',
			fixed: 'a { & ~ /* foo */ a {} }',
			description: 'comment in selector',
			message: messages.expected('explicit'),
			line: 1,
			column: 5,
			endLine: 1,
			endColumn: 18,
			fix: {
				range: [4, 5],
				text: '& ~',
			},
		},
		{
			code: 'a { > .foo, /* bar */ + .baz {} }',
			fixed: 'a { & > .foo, /* bar */ & + .baz {} }',
			description: 'selector list with comments ',
			warnings: [
				{
					message: messages.expected('explicit'),
					line: 1,
					column: 5,
					endLine: 1,
					endColumn: 11,
					fix: {
						range: [4, 5],
						text: '& >',
					},
				},
				{
					message: messages.expected('explicit'),
					line: 1,
					column: 12,
					endLine: 1,
					endColumn: 28,
				},
			],
		},
		{
			code: stripIndent`
				a {
					>  .foo {}
					+
					.bar {}
				}
			`,
			fixed: stripIndent`
				a {
					& >  .foo {}
					& +
					.bar {}
				}
			`,
			description: 'extra spacing in selector',
			warnings: [
				{
					message: messages.expected('explicit'),
					line: 2,
					column: 2,
					endLine: 2,
					endColumn: 9,
					fix: {
						range: [5, 6],
						text: '& >',
					},
				},
				{
					message: messages.expected('explicit'),
					line: 3,
					column: 2,
					endLine: 4,
					endColumn: 6,
					fix: {
						range: [17, 18],
						text: '& +',
					},
				},
			],
		},
	],
});

testRule({
	ruleName,
	config: ['implicit'],
	fix: true,
	computeEditInfo: true,

	accept: [
		{
			code: 'a {}',
			description: 'non-nested rule',
		},
		{
			code: 'a { a {} }',
			description: 'nested rule with implied descendant combinator',
		},
		{
			code: 'a { > a {} }',
			description: 'implicit nesting with child combinator',
		},
		{
			code: 'a { &:hover {} }',
			description: 'nesting with pseudo-class (& required)',
		},
		{
			code: 'a { &.foo {} }',
			description: 'nesting with class (& required)',
		},
		{
			code: 'a { :not(&) {} }',
			description: 'nesting in pseudo-class',
		},
		{
			code: 'a { [foo] {} }',
			description: 'nesting in attribute selector',
		},
		{
			code: 'a { a & {} }',
			description: 'nesting selector as parent',
		},
		{
			code: 'a { + a & {} }',
			description: 'nesting selector as parent and implied initial &',
		},
		{
			code: 'a { a, .foo /* bar */ {} }',
			description: 'nesting selector in list with comment',
		},
		{
			code: 'a { /* bar */ a, /* baz */ .foo {} }',
			description: 'nesting selector in list with starting comments',
		},
	],

	reject: [
		{
			code: 'a { & a {} }',
			fixed: 'a { a {} }',
			message: messages.expected('implicit'),
			line: 1,
			column: 5,
			endLine: 1,
			endColumn: 8,
			fix: {
				range: [4, 6],
				text: '',
			},
		},
		{
			code: 'a { & > a {} }',
			fixed: 'a { > a {} }',
			message: messages.expected('implicit'),
			line: 1,
			column: 5,
			endLine: 1,
			endColumn: 10,
			fix: {
				range: [4, 6],
				text: '',
			},
		},
		{
			code: 'a { & > a & {} }',
			fixed: 'a { > a & {} }',
			message: messages.expected('implicit'),
			line: 1,
			column: 5,
			endLine: 1,
			endColumn: 12,
			fix: {
				range: [4, 6],
				text: '',
			},
		},
		{
			code: 'a { & .foo, /* bar */ & .baz {} }',
			fixed: 'a { .foo, /* bar */ .baz {} }',
			warnings: [
				{
					message: messages.expected('implicit'),
					line: 1,
					column: 5,
					endLine: 1,
					endColumn: 11,
					fix: {
						range: [4, 6],
						text: '',
					},
				},
				{
					message: messages.expected('implicit'),
					line: 1,
					column: 12,
					endLine: 1,
					endColumn: 28,
				},
			],
		},
		{
			code: stripIndent`
				a {
					& .foo {}
					& .bar {}
				}
			`,
			fixed: stripIndent`
				a {
					.foo {}
					.bar {}
				}
			`,
			warnings: [
				{
					message: messages.expected('implicit'),
					line: 2,
					column: 2,
					endLine: 2,
					endColumn: 8,
					fix: {
						range: [5, 7],
						text: '',
					},
				},
				{
					message: messages.expected('implicit'),
					line: 3,
					column: 2,
					endLine: 3,
					endColumn: 8,
					fix: {
						range: [16, 18],
						text: '',
					},
				},
			],
		},
	],
});
```
</details>

**Comment by jeddy3** [comment]
I've pushed some commits to:

- use the PostCSS Selector parser so that we can maintain comments and whitespace
- adhere to our latest conventions
- addressed the remaining review comments

File: lib/rules/relative-selector-nesting-notation/index.mjs [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+			if (!isStandardSyntaxRule(ruleNode)) return;
+
+			if (!isNestedStyleRule(ruleNode)) return;
+
+			const selectorRoot = parseSelector(getRuleSelector(ruleNode), result, ruleNode);
+
+			if (!selectorRoot) return;
+
+			selectorRoot.each((selector) => {
+				if (selector.type !== 'selector') return;
```
Comment (ybiquitous) [inline lib/rules/relative-selector-nesting-notation/index.mjs]: [suggestion] How about using the type guards of `postcss-selector-parser` like this?

```suggestion
				if (!parser.isSelector(selector)) return;
```

File: lib/rules/relative-selector-nesting-notation/index.mjs [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+/**
+ * Check if a selector contains the nesting selector
+ *
+ * @param {import('postcss-selector-parser').Selector} selector
+ * @returns {boolean}
+ */
+function containsNestingSelector(selector) {
+	let hasNestingSelector = false;
+
+	selector.walkNesting(() => (hasNestingSelector = true));
```
Comment (ybiquitous) [inline lib/rules/relative-selector-nesting-notation/index.mjs]: [suggestion] What about halting the loop once the node is found to reduce extra iterations?

```suggestion
	selector.walkNesting(() => {
		hasNestingSelector = true;
		return false; // halt
	});
```

File: lib/rules/relative-selector-nesting-notation/index.mjs [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+ *
+ * @param {import('postcss-selector-parser').Selector} selector
+ * @returns {boolean}
+ */
+function canUseImplicitNesting(selector) {
+	const foundNode = selector.nodes.find(
+		(node) => node.type !== 'comment' && node.type !== 'nesting',
+	);
+
+	return foundNode?.type === 'combinator';
```
Comment (ybiquitous) [inline lib/rules/relative-selector-nesting-notation/index.mjs]: [suggestion] The type guard is also available here (also other places).

```suggestion
	return parser.isCombinator(foundNode);
```

File: lib/rules/relative-selector-nesting-notation/index.mjs [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+
+/**
+ * Fix selector to use implicit nesting notation
+ *
+ * @param {import('postcss-selector-parser').Selector} selector
+ * @param {import('postcss').Rule} ruleNode
+ * @param {import('postcss-selector-parser').Root} selectorRoot
+ */
+function fixImplicit(selector, ruleNode, selectorRoot) {
+	const firstNestingNode = selector.nodes.find((node) => node.type === 'nesting');
```
Comment (ybiquitous) [inline lib/rules/relative-selector-nesting-notation/index.mjs]: [suggestion] Also here. It's simpler and can reduce function creations.

```suggestion
	const firstNestingNode = selector.nodes.find(parser.isNesting);
```

**Comment by jeddy3** [comment]
@ybiquitous Thank you for the review. I've committed all the suggestions.

(I've made a note for myself about the type guards as I often forget about them 😓.)

File: lib/rules/relative-selector-nesting-notation/index.mjs [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+/**
+ * Check if a selector starts with a combinator
+ *
+ * @param {import('postcss-selector-parser').Selector} selector
+ * @returns {boolean}
+ */
+function startsWithCombinator(selector) {
+	const firstSignificantNode = selector.nodes.find((node) => node.type !== 'comment');
```
Comment (ybiquitous) [inline lib/rules/relative-selector-nesting-notation/index.mjs]:
```suggestion
	const firstSignificantNode = selector.nodes.find((node) => !isComment(node));
```

File: lib/rules/relative-selector-nesting-notation/index.mjs [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+
+	return firstSignificantNode?.type === 'combinator';
```
Comment (ybiquitous) [inline lib/rules/relative-selector-nesting-notation/index.mjs]:
```suggestion
	return isCombinator(firstSignificantNode);
```

File: lib/rules/relative-selector-nesting-notation/index.mjs [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+ * Check if a selector can use implicit nesting, i.e.
+ * has a combinator after any comments or nesting selectors
+ *
+ * @param {import('postcss-selector-parser').Selector} selector
+ * @returns {boolean}
+ */
+function canUseImplicitNesting(selector) {
+	const foundNode = selector.nodes.find(
+		(node) => node.type !== 'comment' && node.type !== 'nesting',
+	);
```
Comment (ybiquitous) [inline lib/rules/relative-selector-nesting-notation/index.mjs]:
```suggestion
	const foundNode = selector.nodes.find((node) => !isComment(node) && !isNesting(node));
```

File: lib/rules/relative-selector-nesting-notation/index.mjs [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+
+/**
+ * Fix selector to use explicit nesting notation
+ *
+ * @param {import('postcss-selector-parser').Selector} selector
+ * @param {import('postcss').Rule} ruleNode
+ * @param {import('postcss-selector-parser').Root} selectorRoot
+ */
+function fixExplicit(selector, ruleNode, selectorRoot) {
+	const firstSignificantNode = selector.nodes.find((node) => node.type !== 'comment');
```
Comment (ybiquitous) [inline lib/rules/relative-selector-nesting-notation/index.mjs]:
```suggestion
	const firstSignificantNode = selector.nodes.find((node) => !isComment(node));
```

File: lib/rules/relative-selector-nesting-notation/index.mjs [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+	const firstNestingNode = selector.nodes.find(isNesting);
+	let spaceBeforeFirstNestingNode = '';
+
+	// Remove the first nesting node and store its spaces.before
+	if (firstNestingNode) {
+		spaceBeforeFirstNestingNode = firstNestingNode.spaces.before;
+		firstNestingNode.remove();
+	}
+
+	const firstCombinatorNode = selector.nodes.find((node) => node.type === 'combinator');
```
Comment (ybiquitous) [inline lib/rules/relative-selector-nesting-notation/index.mjs]:
```suggestion
	const firstCombinatorNode = selector.nodes.find(isCombinator);
```

**Comment by jeddy3** [comment]
> I think it's ready to merge now. 👍🏼

Fantastic. Let's hold off on merging the rules for now so we can group them and decide whether they belong in our configs.

Let's shoot for getting the following into the next release instead:
- https://github.com/stylelint/stylelint/pull/9179
- https://github.com/stylelint/stylelint/pull/9199
- https://github.com/stylelint/stylelint/pull/8702

**Comment by sw1tch3roo** [comment]
Wow, thanks for your help with my PRs. Now that I'm free, I think we can continue to implement other features. 👀
